x-pack/plugins/security/server/authorization/privileges/feature_privilege_builder/alerting.ts

```diff
@@ -10,33 +10,52 @@ import { uniq } from 'lodash';
 import type { FeatureKibanaPrivileges, KibanaFeature } from '../../../../../features/server';
 import { BaseFeaturePrivilegeBuilder } from './feature_privilege_builder';
 
-const readOperations: string[] = ['get', 'getAlertState', 'getAlertInstanceSummary', 'find'];
-const writeOperations: string[] = [
-  'create',
-  'delete',
-  'update',
-  'updateApiKey',
-  'enable',
-  'disable',
-  'muteAll',
-  'unmuteAll',
-  'muteInstance',
-  'unmuteInstance',
-];
-const allOperations: string[] = [...readOperations, ...writeOperations];
+enum AlertingType {
+  RULE = 'rule',
+  ALERT = 'alert',
+}
+
+const readOperations: Record<AlertingType, string[]> = {
+  rule: ['get', 'getAlertState', 'getAlertInstanceSummary', 'find'],
+  alert: ['get', 'find'],
+};
+
+const writeOperations: Record<AlertingType, string[]> = {
+  rule: [
+    'create',
+    'delete',
+    'update',
+    'updateApiKey',
+    'enable',
+    'disable',
+    'muteAll',
+    'unmuteAll',
+    'muteInstance',
+    'unmuteInstance',
+  ],
+  alert: ['update'],
+};
+const allOperations: Record<AlertingType, string[]> = {
+  rule: [...readOperations.rule, ...writeOperations.rule],
+  alert: [...readOperations.alert, ...writeOperations.alert],
+};
 
 export class FeaturePrivilegeAlertingBuilder extends BaseFeaturePrivilegeBuilder {
   public getActions(
     privilegeDefinition: FeatureKibanaPrivileges,
     feature: KibanaFeature
   ): string[] {
     const getAlertingPrivilege = (
-      operations: string[],
+      operations: Record<AlertingType, string[]>,
       privilegedTypes: readonly string[],
       consumer: string
     ) =>
-      privilegedTypes.flatMap((type) =>
-        operations.map((operation) => this.actions.alerting.get(type, consumer, operation))
+      privilegedTypes.flatMap((privilegedType) =>
+        Object.values(AlertingType).flatMap((alertingType) =>
+          operations[alertingType].map((operation) =>
+            this.actions.alerting.get(privilegedType, consumer, alertingType, operation)
+          )
+        )
       );
```

Comment:
@gmmorris Is this what we would need to update in order to support sub-feature privileges? For example this scenario:

![image](https://user-images.githubusercontent.com/10927944/116760750-a4e6a900-a9ca-11eb-94fa-8c9ede6921d1.png)

Reply:
TBH I don't know :)
cc @legrego is the person to check with.

Reply:
The `FeaturePrivilegeIterator` has a [`mergeWithSubFeatures` function](https://github.com/elastic/kibana/blob/6ecffcc57c9a5b0328257cd25b737140a8471dda/x-pack/plugins/security/server/authorization/privileges/feature_privilege_iterator/feature_privilege_iterator.ts#L42) that is generally responsible for this support.


There is already an `alerting` section within this function, so as long as the semantics of alerting haven't changed within the `features` plugin, then I _expect_ that you'll be all set. It's easy enough for us to test when we review this PR

https://github.com/elastic/kibana/blob/6ecffcc57c9a5b0328257cd25b737140a8471dda/x-pack/plugins/security/server/authorization/privileges/feature_privilege_iterator/feature_privilege_iterator.ts#L84-L90

Reply:
Are we going to just use `alerting.all` and `alerting.read` for now?
I guess we can split them into  `alerting.rules.all` and `alerting.alerts.all` later if we need to diverge, right?

Reply:
Yeah from my perspective we can choose to diverge them at any point

Reply:
:partywallaby:

Reply:
> Are we going to just use `alerting.all` and `alerting.read` for now?
> I guess we can split them into `alerting.rules.all` and `alerting.alerts.all` later if we need to diverge, right?

@ymao1 and I met to discuss this today, and after she helped me understand the needs, we came to the conclusion that we will need to split this out into `alerting.rules.*` and `alerting.alerts.*` as you suggested, @gmmorris. We are thinking that this work should be done in its own PR though, as it'll be a fairly widespread change

Reply:
Issue for subfeature privilege here: https://github.com/elastic/kibana/issues/99940

Reply:
SGTM 👍